# Snakemake workflow: Snakemake CRISPR Barcode/HiTom Editing Analysis

![pngwing com](https://github.com/user-attachments/assets/17f90df0-c033-4b04-883f-6d8f981066ba)

A snakemake workflow for CRISPR barcode designer and analysis of Illimina sequences pair-end reads (based on [CRISPResso2](https://github.com/pinellolab/CRISPResso2)).

[![Snakemake](https://img.shields.io/badge/snakemake-â‰¥5.2.1-brightgreen.svg)](https://snakemake.bitbucket.io)
[![Build Status](https://travis-ci.org/snakemake-workflows/rna-seq-star-deseq2.svg?branch=master)](https://travis-ci.org/snakemake-workflows/rna-seq-star-deseq2)
[![Snakemake-Report](https://img.shields.io/badge/snakemake-report-green.svg)](https://cdn.rawgit.com/snakemake-workflows/rna-seq-star-deseq2/master/.test/report.html)


## Authors

* Zhongping Xu (@hopetogy), http://tiramisutes.github.io/

## Usage

---

è¯·ä»”ç»†é˜…è¯» **config.yaml** æ–‡ä»¶ä¸­å‚æ•°çš„æ³¨é‡Šè¯´æ˜ï¼Œä¸¥æ ¼æ­£ç¡®çš„é…ç½®ç›¸åº”å‚æ•°ï¼Œç‰¹åˆ«æ˜¯**prefix**å’Œ**DATA**çš„å‚æ•°ã€‚

é…ç½®æ–‡ä»¶å†…å®¹è¯·å‚è€ƒ**Resources**ç›®å½•æä¾›çš„testæ•°æ®è¿›è¡Œå‡†å¤‡ã€‚

åˆ‡è®°ï¼šè¾“å…¥çš„ä»»ä½•æ–‡ä»¶å†…å®¹æˆ–æ–‡ä»¶åç§°å¿…é¡»å‡ä¸å«ç‰¹æ®Šå­—ç¬¦ï¼Œä»…å¯ä¸ºå­—æ¯ã€æ•°å­—ã€ä¸­çº¿ (-)ã€ç‚¹ (ä»…å¯å‡ºç°åœ¨æ–‡ä»¶åç§°ä¸­)å’Œä¸‹åˆ’çº¿ (_)ã€‚

---

### Step 1: Install workflow / git clone

First install [Snakemake](https://snakemake.readthedocs.io/en/stable/getting_started/installation.html):

```
conda create -n snakemake -c conda-forge -c bioconda snakemake conda-forge::colorama
conda activate snakemake
```

Then git clone this repository using follows command ğŸ‘‡

    git clone https://github.com/tiramisutes/Snakemake_CRISPR_Barcode_HiTom_Analysis.git

### Step 2: Prepare data
- Hi-Tomè¿”å›æ•°æ®å·²ç»æŒ‰ç…§æ ·å“è¿›è¡Œfqæ–‡ä»¶çš„åˆ†éš”ï¼Œæ‰€ä»¥åªéœ€æŒ‡å®šæ ·å“åç§°å¯¹åº”çš„æ‰©å¢åºåˆ—å’ŒsgRNAåºåˆ—å³å¯ï¼›
- Barcodeæµ‹åºè¿”å›çš„æ˜¯æ‰€æœ‰æ ·å“çš„æ··åˆç»“æœï¼Œéœ€è¦æ ¹æ®Barcodeå’Œå¼•ç‰©åºåˆ—å°†æ ·å“åˆ†å‰²ï¼Œæ‰€ä»¥éœ€è¦å‡†å¤‡ä¸¤ä¸ªæ–‡ä»¶ï¼šå¼•ç‰©å’Œæ‰©å¢åºåˆ—ä¿¡æ¯ï¼Œä¸”ä¸¤ä¸ªæ–‡ä»¶ä¸­çš„æ ·å“åç§°å¿…é¡»å®Œå…¨ä¸€è‡´ï¼›

|                 | HiTom  | Barcode |
| :-------------: | :----: | ------- |
|   æ‰©å¢å¼•ç‰©ä¿¡æ¯   | ä¸éœ€è¦  | éœ€è¦    |
| æ‰©å¢åºåˆ—å’ŒsgRNA  |  éœ€è¦   | éœ€è¦    |

#### æ‰©å¢åºåˆ—å’Œå¯¹åº”sgRNAæ–‡ä»¶
Tabåˆ†éš”ä¸”å„åˆ—ä¾æ¬¡ä¸ºï¼š`æ‰©å¢åç§°ï¼Œæ‰©å¢åºåˆ—1ï¼Œé¶æ ‡1ï¼Œæ‰©å¢åºåˆ—2ï¼ˆå¯é€‰ï¼‰ï¼Œé¶æ ‡2ï¼ˆå¯é€‰ï¼‰`ï¼›

åŒä¸€æ‰©å¢åºåˆ—å†…æœ‰å¤šä¸ªsgRNAæ—¶ï¼Œåœ¨ç›¸åº”åˆ—ç”¨é€—å·åˆ†éš”ï¼Œå‚è€ƒæ ¼å¼å¦‚ä¸‹ğŸ‘‡

```
312	CATGAATTTTTTCCATCTGCAGTTACTAACAAGCCTAAGGTTGAGGTTCATGGAGGTGATATGAGATCCTTTTTCACCCTGGTAACTATACATGTCGCTAATACTCTCTTTTCTTTTCTTTTT	CCTAAGGTTGAGGTTCATGG	TTGATTAAGAGAAGTACCAGTGTAAGTGCTCCCTCAGGTAAGGGTCACTAGGACCAGGAACATCTGGGTCTGTCATCACCTGCAATATAAAAATATGATTGCTGATAGACATTTTCTCTTGAA	TCAGGTAAGGGTCACTAGGA
313	CATGAATTTTTTCCATCTGCAGTTACTAACAAGCCTAAGGTTGAGGTTCATGGAGGTGATATGAGATCCTTTTTCACCCTGGTAACTATACATGTCGCTAATACTCTCTTTTCTTTTCTTTTT	CCTAAGGTTGAGGTTCATGG,CCTAAGGTTGAGGTTCATGG
```
#### å¼•ç‰©ä¿¡æ¯æ–‡ä»¶
```
312-F	acagggTTTTACCCTTACTTATAGTTATAAA
312-R	ccctgtATGTAACCGAAGAAAGCAT
313-F	accatcTTTTACCCTTACTTATAGTTATAAA
313-R	gatggtATGTAACCGAAGAAAGCAT
```
#### HiTomæ•°æ®
> Hi-Tomæµ‹åºä¼šåœ¨è‡ªå·±è®¾è®¡å¼•ç‰©å‰åŠ 27bpçš„æµ‹åºæ¥å¤´+barcodeåºåˆ—ï¼Œæ€»å…±å°±æ˜¯54bpï¼Œè€Œæµ‹åºæ€»é•¿åº¦ä¸º `150 X 2 = 300bp`ï¼Œä¸”æœ€å¥½å·¦å³åŒç«¯åºåˆ—æœ‰10bpçš„é‡å ä»¥ä¾¿äºåˆå¹¶ï¼ˆflashé»˜è®¤æŒ‰ç…§10bpé‡å è¿›è¡Œåˆå¹¶ï¼‰ï¼Œæ‰€ä»¥æ‰©å¢åºåˆ—æ€»é•¿åº¦æœ€å¥½å°äºç­‰äº236 bpã€‚

å…¬å¸è¿”å›fqæµ‹åºæ•°æ®å‹ç¼©æ–‡ä»¶ï¼Œä¸”æµ‹åºæ•°æ®æ–‡ä»¶å¿…é¡»æŒ‰ç…§å¦‚ä¸‹å‘½åè§„ç¨‹ï¼š`prefixå‰ç¼€-æ ·æœ¬åç§°-R1.fq`ï¼Œä¾‹å¦‚SY207-USR-633-340-R1.fqå’ŒSY207-USR-633-340-R2.fq

- æ ·å“åç§°å¿…é¡»ä¸ä¸‹é¢descriptionæ–‡ä»¶ä¸­çš„ç¬¬ä¸€åˆ—åç§°å®Œå…¨ä¸€è‡´
- å¿…é¡»æ˜¯`R1`å’Œ`R2`ï¼Œä¸”æ˜¯`-`è¿æ¥

### Step 3: Configure workflow

Configure the workflow according to your needs via editing the file `config.yaml`.

### Step 4: Execute workflow

Test your configuration by performing a dry-run via

    snakemake --use-conda -n

Execute the workflow locally via

    snakemake --use-conda --cores $N

using `$N` cores or run it in a cluster environment via

    snakemake --use-conda --cluster qsub --jobs 100

or

`cluster.json`æ–‡ä»¶è¿›è¡Œé›†ç¾¤è¿è¡Œç›¸å…³å‚æ•°è®¾ç½®ï¼Œå› ä¸º`CRISPResso2` ç¨‹åºå¯¹æœåŠ¡å™¨è¿è¡Œæ²¡æœ‰ç‰¹æ®Šè¦æ±‚ï¼Œé€šå¸¸é»˜è®¤å³å¯ï¼›

    snakemake --use-conda --jobs 100 --cluster-config config/cluster.json --cluster "bsub -q {cluster.queue} -o {cluster.output} -e {cluster.error}"

See the [Snakemake documentation](https://snakemake.readthedocs.io/en/stable/executable.html) for further details.

### Step 5: Investigate results

After successful execution, you can create a self-contained interactive HTML report with all results via:

    snakemake --report report.html --report-stylesheet config/custom-stylesheet.css

This report can, e.g., be forwarded to your collaborators.
An example (using some trivial test data) can be seen [here](https://cdn.rawgit.com/snakemake-workflows/rna-seq-star-deseq2/master/.test/report.html).

## FAQ
### 1. ERROR: Error: No alignments were found
å³æ²¡æœ‰æµ‹åºreadsæ¯”å¯¹åˆ°æ‰©å¢åºåˆ—ï¼Œé¦–å…ˆç¡®è®¤æ‰©å¢åºåˆ—æ€»é•¿åº¦æ˜¯å¦å°äºç­‰äº236 bpã€‚

æ‰‹åŠ¨è¿è¡Œ `flash SY207-USR-633-321-1.fastq.gz SY207-USR-633-321-2.fastq.gz --min-overlap 10`å‘½ä»¤ï¼Œå¯Ÿçœ‹ **out.extendedFrags.fastq** æ–‡ä»¶ç¡®è®¤é‡å readsæ•°ç›®ã€‚

è‹¥**out.extendedFrags.fastq** é‡å readsæ•°ç›®å°‘ï¼Œå¯è®¾ç½® `--min_paired_end_reads_overlap` å‚æ•°è¿›è¡Œä¸€å®šçš„æ‹¯æ•‘ã€‚å¦‚æœè¿˜æ˜¯ä¸è¡Œï¼Œå‚è€ƒ[3. å¦‚ä½•æŒ½æ•‘æ‰©å¢åºåˆ—æ€»é•¿åº¦å¤§äº236 bpçš„ç»“æœ](#3-å¦‚ä½•æŒ½æ•‘æ‰©å¢åºåˆ—æ€»é•¿åº¦å¤§äº236-bpçš„ç»“æœ)ã€‚

è‹¥**out.extendedFrags.fastq** é‡å readsæ•°ç›®è¶³å¤Ÿå¤šï¼ŒCRISPResso`--default_min_aln_score`é»˜è®¤60%ç›¸ä¼¼æ€§å³è®¤ä¸ºæµ‹åºreadsæ¯”å¯¹åˆ°æ‰©å¢åºåˆ—ï¼Œæ­¤æ—¶å¯å¯Ÿçœ‹ç¡®è®¤æ˜¯å¦æµ‹åºreaså‰é¢æœ‰å¤ªå¤šçš„æ¥å¤´åºåˆ—ï¼Œé€ æˆæ¯”å¯¹ç›¸ä¼¼å€¼å°äº60%ï¼Œç‰¹åˆ«æ˜¯å•ç«¯readsæ¨¡å¼è·‘å®¹æ˜“å‡ºç°è¿™ç§æƒ…å†µã€‚

### 2. å›¾9ç»“æœreadséå¸¸å°‘
![](workflow/report/9.Alleles_frequency_table_around_sgRNA_TCCTAGTGACCCTTACCTGA.png)
ä¸»è¦åŸå› è¿˜æ˜¯æ¯”å¯¹ä¸Šçš„readså¤ªå°‘ã€‚å¯å‚è€ƒ [`ERROR: Error: No alignments were found`](#1-error-error-no-alignments-were-found) é—®é¢˜è¿›è¡Œæ£€æŸ¥ã€‚
### 3. å¦‚ä½•æŒ½æ•‘æ‰©å¢åºåˆ—æ€»é•¿åº¦å¤§äº236 bpçš„ç»“æœ
å¦‚æœæ˜¯åŒé¶æ ‡ä¸”é¶æ ‡ä½ç½®åˆ†åˆ«åœ¨åŒç«¯readsçš„ä¸¤ç«¯åºåˆ—ï¼Œåˆ™åˆ†åˆ«æŒ‰ç…§å•é¶æ ‡è·‘ï¼Œæ­¤æ—¶åœ¨`config.yaml`ä¸­è®¾ç½®å‚æ•°`badreads: true`å³å¯ã€‚

å¦‚æœæ˜¯å•é¶æ ‡åˆ™ç”¨ä¿®æ”¹åçš„æ‰©å¢åºåˆ—è·‘ç›¸åº”readsã€‚

## åŸºå› ç¼–è¾‘æ£€æµ‹
![](docs/gbp1.PNG)
![](docs/gbp2.PNG)
![](docs/gbp3.PNG)
